crates/revm/src/state_change.rs

```diff
@@ -51,11 +56,127 @@ pub fn post_block_balance_increments(
+            ..Default::default()
+        })
+        .into();
+
+    // Insert the state change for the slot
+    let (slot, value) = eip2935_block_hash_slot(block_number - 1, db)
+        .map_err(|err| BlockValidationError::Eip2935StateTransition { message: err.to_string() })?;
+    account.storage.insert(slot, value);
+
+    // If the first slot in the ring is `U256::ZERO`, then we can assume the ring has not been
+    // filled before, and this is the activation block.
+    //
+    // Reasoning:
+    // - If `block_number <= HISTORY_SERVE_WINDOW`, then the ring will be filled with as many blocks
+    //   as possible, down to slot 0.
+    //
+    //   For example, if it is activated at block 100, then slots `0..100` will be filled.
+    //
+    // - If the fork is activated at genesis, then this will only run at block 1, which will fill
+    //   the ring with the hash of block 0 at slot 0.
+    //
+    // - If the activation block is above `HISTORY_SERVE_WINDOW`, then `0..HISTORY_SERVE_WINDOW`
+    //   will be filled.
+    let is_activation_block = db
+        .storage(HISTORY_STORAGE_ADDRESS, U256::ZERO)
+        .map_err(|err| BlockValidationError::Eip2935StateTransition { message: err.to_string() })?
+        .is_zero();
+
+    // If this is the activation block, then we backfill the storage of the account with up to
+    // `HISTORY_SERVE_WINDOW - 1` ancestors' blockhashes as well, per the EIP.
+    //
+    // Note: The -1 is because the ancestor itself was already inserted up above.
+    if is_activation_block {
+        let mut ancestor_block_number = block_number - 1;
+        for _ in 0..HISTORY_SERVE_WINDOW - 1 {
+            // Stop at genesis
+            if ancestor_block_number == 0 {
+                break
+            }
+            ancestor_block_number -= 1;
+
+            let (slot, value) =
+                eip2935_block_hash_slot(ancestor_block_number, db).map_err(|err| {
+                    BlockValidationError::Eip2935StateTransition { message: err.to_string() }
+                })?;
+            account.storage.insert(slot, value);
+        }
+    }
+
+    // Mark the account as touched and commit the state change
+    account.mark_touch();
+    db.commit(HashMap::from([(HISTORY_STORAGE_ADDRESS, account)]));
+
+    Ok(())
+}
+
+/// Helper function to create a [`StorageSlot`] for [EIP-2935] state transitions for a given block
+/// number.
+///
+/// This calculates the correct storage slot in the `BLOCKHASH` history storage address, fetches the
+/// blockhash and creates a [`StorageSlot`] with appropriate previous and new values.
+fn eip2935_block_hash_slot<DB: Database>(
+    block_number: u64,
+    db: &mut DB,
+) -> Result<(U256, StorageSlot), DB::Error> {
+    let slot = U256::from(block_number).rem(U256::from(HISTORY_SERVE_WINDOW));
+    let current_hash = db.storage(HISTORY_STORAGE_ADDRESS, slot)?;
+    let ancestor_hash = db.block_hash(U256::from(block_number))?;
```

Comment:
This does not make sense for me to fetch `ancestor_hash`, only thing we care about is what is inside the storage

Reply:
Aha this made sense for backfilling, but with newest changes with just need `header.parent_hash`

Reply:
makes sense